### PR TITLE
big array server restart crash and qalter usage correction

### PR DESCRIPTION
**Pull request by Shrini-h**
#### Describe Bug or Feature

1. Server crashes during restart, while recovering a job array whose size is bigger than the currently set `max_array_size` server attribute.

2. `qalter` usage string wrongly mentions Job array option `[-J X-Y[:Z]]`, which is infact not supported by it.
```
$ qalter -J 1-20 7[].pbsserver
qalter: invalid option -- 'J'
usage: qalter [-a date_time] [-A account_string] [-c interval] [-e path]
        [-h hold_list] [-j y|n] [-k keep] [-J X-Y[:Z]] [-l resource_list]
        [-m mail_options] [-M user_list] [-N jobname] [-o path] [-p priority]
        [-R o|e|oe] [-r y|n] [-S path] [-u user_list] [-W dependency_list]
        [-P project_name] job_identifier...
       qalter --version
```

3. Invalid read of 4 bytes with below stack while executing TestJobArray PTL testsuite
```
==64194== Invalid read of size 4
==64194==    at 0x47B31F: req_deletejob (req_delete.c:445)
==64194==    by 0x4779CA: dispatch_request (process_request.c:683)
==64194==    by 0x477666: process_request (process_request.c:488)
==64194==    by 0x4F117D: process_socket (net_server.c:504)
==64194==    by 0x4F148C: wait_request (net_server.c:620)
==64194==    by 0x475BC3: main (pbsd_main.c:2141)
```

4. Dead code found functions: `node_recov_db()` and `db_to_svr_node()` in `node_recov_db.c`

#### Describe Your Change

1. made `decode_attr_db()` to respond to the return code of `at_action()` action functions of attributes
2. made `mk_subjob_index_tbl()` to compare Job array size with `max_array_size` server attribute only for `ATR_ACTION_NEW` and `ATR_ACTION_ALTER`
3. added validation of job array flags at the beginning of action function of `array_indices_remaining`, so that we can avoid server crash
4. corrected `qalter`'s usage by removing Job array option  `[-J X-Y[:Z]]`
5. Fixed invalid read by moving the offending lines to appropriate locations
6. purged the dead code functions: `node_recov_db()` and `db_to_svr_node()` in `node_recov_db.c`

#### Attach Test and Valgrind Logs/Output

[before-fail.zip](https://github.com/PBSPro/pbspro/files/3637418/before-fail.zip)
[after-pass.zip](https://github.com/PBSPro/pbspro/files/3637420/after-pass.zip)
[SmokeTest,TestJobArray.zip](https://github.com/PBSPro/pbspro/files/3637419/SmokeTest.TestJobArray.zip)

[valgrind_invalid_fixed.log](https://github.com/PBSPro/pbspro/files/3646867/valgrind_invalid_fixed.log)
[PTL_pass_report_python3.zip](https://github.com/PBSPro/pbspro/files/3647883/PTL_pass_report_python3.zip)

PTL report after commit https://github.com/PBSPro/pbspro/commit/18202d3f6d55f4a408c9bc5083f63470f9649d78
[PTL_pass_report_18202d3.zip](https://github.com/PBSPro/pbspro/files/3706605/PTL_pass_report_18202d3.zip)

### CODE REVIEW

**Comment by prakashcv13** [comment]
restarted the travis opensuse/leap:15 build

**Comment by Shrini-h** [comment]
While executing valgrind, I found below invalid read, which I will be fixing in the next commit
```
==64194== Invalid read of size 4
==64194==    at 0x47B31F: req_deletejob (req_delete.c:445)
==64194==    by 0x4779CA: dispatch_request (process_request.c:683)
==64194==    by 0x477666: process_request (process_request.c:488)
==64194==    by 0x4F117D: process_socket (net_server.c:504)
==64194==    by 0x4F148C: wait_request (net_server.c:620)
==64194==    by 0x475BC3: main (pbsd_main.c:2141)
==64194==  Address 0x14a87b80 is 176 bytes inside a block of size 6,448 free'd
==64194==    at 0x4C2ACBD: free (vg_replace_malloc.c:530)
==64194==    by 0x454294: job_free (job_func.c:599)
==64194==    by 0x454939: job_purge (job_func.c:1001)
==64194==    by 0x4BF352: svr_saveorpurge_finjobhist (svr_jobfunc.c:5050)
==64194==    by 0x47C162: req_deletejob2 (req_delete.c:823)
==64194==    by 0x47B31A: req_deletejob (req_delete.c:444)
==64194==    by 0x4779CA: dispatch_request (process_request.c:683)
==64194==    by 0x477666: process_request (process_request.c:488)
==64194==    by 0x4F117D: process_socket (net_server.c:504)
==64194==    by 0x4F148C: wait_request (net_server.c:620)
==64194==    by 0x475BC3: main (pbsd_main.c:2141)
==64194==  Block was alloc'd at
==64194==    at 0x4C29BC3: malloc (vg_replace_malloc.c:299)
==64194==    by 0x453D1F: job_alloc (job_func.c:321)
==64194==    by 0x43CA54: create_subjob (array_func.c:862)
==64194==    by 0x49AF90: req_runjob (req_runjob.c:522)
==64194==    by 0x477BA8: dispatch_request (process_request.c:773)
==64194==    by 0x477666: process_request (process_request.c:488)
==64194==    by 0x4F117D: process_socket (net_server.c:504)
==64194==    by 0x4F12F8: wait_request (net_server.c:585)
==64194==    by 0x475BC3: main (pbsd_main.c:2141)
==64194== 
```

**Comment by prakashcv13** [comment]
hey @Shrini-h, if you are still working on this, should we mark this PR WIP?

**Comment by Shrini-h** [comment]
> hey @Shrini-h, if you are still working on this, should we mark this PR WIP?

As you wish, hope rebasing to python 3 commit will go smooth and quick

**Comment by Shrini-h** [comment]
PR ready for review/merge

**Comment by prakashcv13** [comment]
@Shrini-h - The issue, if I understand correctly, will happen if max_array_size is reduced to a value less than the size of an array job already submitted.
instead of skipping the check in mk_subjob_index_tbl, should we make a change to not allow max_array_size be reduced less than the size of an array job already submitted. We would need the server to keep track of the biggest array job submitted so far.
Also, I am not able to see much utility in adding a new parameter to decode_attr_db() just for a new log message.

**Comment by Shrini-h** [comment]
@prakashcv13  thank you for your inputs. Please find my replies below
> instead of skipping the check in mk_subjob_index_tbl, should we make a change to not allow max_array_size be reduced less than the size of an array job already submitted. We would need the server to keep track of the biggest array job submitted so far.

This is a nice idea. However it will severely cripple admins capability to reduce the job array size. He will have to wait for the biggest job array to finish or kill it before reducing the array size. Also we will have to worry about race conditions that can crop up by tracking biggest array job. Hence I do not recommend this method.

> Also, I am not able to see much utility in adding a new parameter to decode_attr_db() just for a new log message.

Just wait till a customer reports missing objects after server restart, This will definitely help admin to track objects which fail to load from db during server restarts.

**Comment by prakashcv13** [comment]
> Just wait till a customer reports missing objects after server restart, This will definitely help admin to track objects which fail to load from db during server restarts.
We haven't got any request from the customer who needs such a log. If you feel it is really necessary, find out if it is possible to do this from the caller of decode_attr_db(). We are adding a parameter only for a log message. If it is not possible to do it from the caller, let us wait for a customer who reports such issue.

**Comment by Shrini-h** [comment]
> We haven't got any request from the customer who needs such a log. If you feel it is really necessary, find out if it is possible to do this from the caller of decode_attr_db(). We are adding a parameter only for a log message. If it is not possible to do it from the caller, let us wait for a customer who reports such issue.

I believe logging should be predictive and appropriately placed so as to be easily helpful in case of failures. It should help the admins / support to figure out easily the failure points without going to and fro because of lack of appropriate logs. If you feel it is unnecessary please justify and also explain if any penalty introduced because of an extra parameter to a non-api function. I will be happy to remove the new parameter and the impugned log statement!

**Comment by prakashcv13** [comment]
> I believe logging should be predictive and appropriately placed so as to be easily helpful in case of failures. It should help the admins / support to figure out easily the failure points without going to and fro because of lack of appropriate logs. If you feel it is unnecessary please justify and also explain if any penalty introduced because of an extra parameter to a non-api function. I will be happy to remove the new parameter and the impugned log statement!

I do not see any penalty, but adding a parameter just for a log message does not seem like a great design to me. So, it is better if we do it in the caller.

**Comment by Shrini-h** [comment]
> I do not see any penalty, but adding a parameter just for a log message does not seem like a great design to me. So, it is better if we do it in the caller.

It might not be a great design in your view, but in my view its a good design. It will also avoid an extra log line and there is added advantage of getting all the necessary info/context on single line. (should make any log parser happy).  I appreciate if you can justify why it "_does not seem like a great design_" to you (for it to be amended). You have already agreed there's no penalty.

**Comment by prakashcv13** [comment]
> It might not be a great design in your view, but in my view its a good design. It will also avoid an extra log line and there is added advantage of getting all the necessary info/context on single line. (should make any log parser happy). I appreciate if you can justify why it "_does not seem like a great design_" to you (for it to be amended). You have already agreed there's no penalty.

well, we clearly have different ideas about adding a new parameter to a function. I do not see much utility/use/purpose in adding a new parameter to a function for the purpose recording a log message in case of a rare failure. And, this log is not *closely* related to the problem we are trying to solve here (again, according to me), and you tend to disagree. I will agree to disagree here and leave the PR open for other reviewers.

**Comment by Shrini-h** [comment]
> well, we clearly have different ideas about adding a new parameter to a function. I do not see much utility/use/purpose in adding a new parameter to a function for the purpose recording
 
Well I tried hard justifying all my decisions and listed out all the advantages, if that doesn't help push my view, I don't know what will. I really failed to extract any logic justification from your arguments even though I explicitly requested for it.

> log message in case of a rare failure.

If you excuse me for delving a bit into information theory: "_The amount of information of an event is inversely proportional to the probability of the event_". So a rare failure has more juicy information and hence its right place to do lot of logging (including object names 😄). I'm sorry if this argument doesn't makes sense but believe me its a universal truth

> And, this log is not _closely_ related to the problem we are trying to solve here (again, according to me), and you tend to disagree.

Well the problem we are trying to solve here is because of lacunae in the design, which i have tried my best to rectify with justification. Targeting the root cause is always better than patching the symptom.

> I will agree to disagree here and leave the PR open for other reviewers.

Yes, Please feel free to disagree. Since there is already a work around, lets keep the PR open for other reviewers.

**Comment by arungrover** [comment]
I think that your intent of logging relevant information on failure is right but passing in an object name may not be the best way to go about it. Potentially two different objects could have the same object name. For instance, I can create a multi-sched for big memory jobs and name it "bigmem" and I can also create a queue with name "bigmem". So even if you log it, you will not know which object was being recovered. 
If attribute recovery fails we bail out anyways from decode_attr_db function. So I agree with Prakash that it is the responsibility of the caller to log the object name and I do see we are already handling it in many places.

**Comment by Shrini-h** [comment]
> I think that your intent of logging relevant information on failure is right but passing in an object name may not be the best way to go about it. Potentially two different objects could have the same object name. For instance, I can create a multi-sched for big memory jobs and name it "bigmem" and I can also create a queue with name "bigmem". So even if you log it, you will not know which object was being recovered.
> If attribute recovery fails we bail out anyways from decode_attr_db function. So I agree with Prakash that it is the responsibility of the caller to log the object name and I do see we are already handling it in many places.

@arungrover Thank you, This is the kind of justification I agree upon.
I will shortly remove the object name parameter and make sure it is logged on all the parents in case of failures.

**Comment by Shrini-h** [comment]
@prakashcv13 @arungrover 
review comment on object name parameter has been addressed in https://github.com/PBSPro/pbspro/pull/1308/commits/18202d3f6d55f4a408c9bc5083f63470f9649d78
Rebased to latest and rerun PTL report is [here](https://github.com/PBSPro/pbspro/files/3706605/PTL_pass_report_18202d3.zip)
Please check it out.
Thank you

File: src/server/node_recov_db.c [hunk 1]
```diff
@@ -373,8 +375,15 @@ node_recov_db(void *nd)
 	return np;
 
 db_err:
+	if (np_init) {
```
Comment (arungrover) [inline src/server/node_recov_db.c]: I don't see this function being called anywhere. I am not sure if this function is even needed anymore. Do you really want to keep this change?

Comment (Shrini-h) [reply src/server/node_recov_db.c]: Thanks for noticing, never realized its so.
I have now purged dead functions node_recov_db() and static db_to_svr_node() in https://github.com/PBSPro/pbspro/pull/1308/commits/8060bef991e0d99022e782e4bd322fa21548b336

Comment (Shrini-h) [reply src/server/node_recov_db.c]: This fact has been updated in the heading comment/ bug description

**Comment by arungrover** [comment]
@Shrini-h I see no one else has picked up this review, you should ask one of your team members to help you out with review here

**Comment by Shrini-h** [comment]
@arungrover , my job is done here, unless there is a new comment. Please feel free to nominate.
Thank you

**Comment by arungrover** [comment]
Thanks @nithinj

**Pull request merged**